File: 0122-best-time-to-buy-and-sell-stock-ii/0122-best-time-to-buy-and-sell-stock-ii.java
class Solution {
    
    int [][]dp;
    public int memo(int[] prices,int i,int buy){
        if(i==prices.length)return 0;
        if(dp[i][buy]!=-1)return dp[i][buy];
        
        if(buy==1){
           return dp[i][buy]=Math.max(-prices[i]+memo(prices,i+1,0),memo(prices,i+1,1));
        }
        return dp[i][buy]=Math.max(prices[i]+memo(prices,i+1,1),memo(prices,i+1,0));
        
    }
    
    public int maxProfit(int[] prices) {
        int n=prices.length;
        dp=new int[n+1][2];
        // for(int []i:dp)
        //     Arrays.fill(i,-1);
        // return memo(prices,0,1);
        
        for(int i=n-1;i>=0;i--){
            for(int buy=1;buy>=0;buy--){
                if(buy==1){
                    dp[i][buy]=Math.max(-prices[i]+dp[i+1][0],dp[i+1][1]);
                }
                else dp[i][buy]=Math.max(prices[i]+dp[i+1][1],dp[i+1][0]);
            }
        }
        return dp[0][1];
    }
}

File: 0122-best-time-to-buy-and-sell-stock-ii/0122-best-time-to-buy-and-sell-stock-ii.cpp
class Solution {
public:
    int dp[30001][2];
    int recur(int idx,vector<int>&prices,bool buy){
        if(idx>=prices.size())return 0;
        if(dp[idx][buy]!=-1)return dp[idx][buy];
        if(buy){
            return dp[idx][buy]= max(prices[idx]+recur(idx+1,prices,!buy),recur(idx+1,prices,buy));
        }
        return dp[idx][buy]= max(-prices[idx]+recur(idx+1,prices,true),recur(idx+1,prices,buy));
        
    }
    

    int maxProfit(vector<int>& prices) {
//         int n=prices.size();
//         vector<vector<int>>dp(n+1,vector<int>(2));
    
//         for(int i=n-1;i>=0;i--){
//             for(int buy=0;buy<=1;buy++){
//                 if(buy)
//                     dp[i][buy]=max(-prices[i]+dp[i+1][0],dp[i+1][1]);
//                 else
//                     dp[i][buy]=max(prices[i]+dp[i+1][1],dp[i+1][0]);
//             }
//         }
//         return dp[0][1];
        memset(dp,-1,sizeof(dp));
        return recur(0,prices,false);
        
    }
};